### PR TITLE
XRT-290 Enable clock subdev for versal

### PR DESCRIPTION
**Pull request by xdavidz**
Pre-review of clock working for versal. Please just take a brief look, some corner cases have not been finalized yet.

I just want to see if pipeline is happy.

Given it is sort of 'complicated', we could have a phone call. 
Here is the spec: https://confluence.xilinx.com/display/XIP/VCK5000+kernel+clocking#VCK5000kernelclocking-ClockScaling

Test Result:
> [root@xsjhemn41 davidz]# xbutil dump|grep clock
            "clock0": "200",
            "clock1": "300",
            "clock2": "0",
[root@xsjhemn41 davidz]# xbutil clock -f 100
INFO: Found total 1 card(s), 1 are usable
INFO: xbutil clock succeeded.
[root@xsjhemn41 davidz]# xbutil dump|grep clock
            "clock0": "100",
            "clock1": "300",
            "clock2": "0",
[root@xsjhemn41 davidz]# xbutil clock -g 150
INFO: Found total 1 card(s), 1 are usable
INFO: xbutil clock succeeded.
[root@xsjhemn41 davidz]# xbutil dump|grep clock
            "clock0": "100",
            "clock1": "150",
            "clock2": "0",
[root@xsjhemn41 davidz]# xbutil clock -f 210
INFO: Found total 1 card(s), 1 are usable
ERROR: xbutil clock failed.
[root@xsjhemn41 davidz]# xbutil clock -g 301
INFO: Found total 1 card(s), 1 are usable
ERROR: xbutil clock failed.
[root@xsjhemn41 davidz]# xbutil clock -g 300
INFO: Found total 1 card(s), 1 are usable
INFO: xbutil clock succeeded.
[root@xsjhemn41 davidz]# xbutil dump|grep clock
            "clock0": "100",
            "clock1": "300",
            "clock2": "0",
[root@xsjhemn41 davidz]#

also tested on u50 and u200 qdma and xdma

### CODE REVIEW

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/axigate.c [hunk 1]
```diff
@@ -73,6 +73,8 @@ static int axigate_free(struct platform_device *pdev)
 	if (freeze)
 		goto done; /* Already free */
 
+	reg_wr(gate, 0x1, iag_wr);
```
[the remark this answers is not in the document]
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/axigate.c:76]: I've asked them to change to 0x2.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c [hunk 2]
```diff
@@ -385,6 +625,165 @@ static unsigned short clock_get_freq_impl(struct clock *clock, int idx)
...
+		clkout0_ts = (struct acap_clkout0_ts *)&val;
+		clkout0_ts->clkout0_lt = (O - (O % 2)) / 2;
+		clkout0_ts->clkout0_ht = (O - (O % 2)) / 2;
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_CLKOUT0_TS, val);
+
+		/* init the freq change */
+		reg_wr(clock->clock_bases[i] + OCL_CLKWIZ_INIT_CONFIG, 0x3);
+		err = clock_wiz_busy(clock, i, 100, 100);
+		if (err) {
+			/* ??? should we restore back if fail */
```
Comment (larry9523) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c:778]: Yeah, seems we already did in line 859? Or at least we should log a better error message?

**Comment by xdavidz** [comment]
retest this please

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c (hunk 2 above)
Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c:778]: There is no equivalent rollback procedure for ACAP Versal yet, the best way here might be just bail out with waning message from clock_wiz_busy.

**Comment by xdavidz** [comment]
@houlz0507 @larry9523 I have removed the "do not merge" label, please let me know if you need a coding review meeting. I can go over the changes for you.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c (hunk 2 above)
Comment (larry9523) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c]: OK, fair enough

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c [hunk 3]
```diff
@@ -460,16 +856,9 @@ static int clock_ocl_freqscaling_impl(struct clock *clock, bool force,
...
-			val = reg_rd(clock->clock_bases[i] +
-				OCL_CLKWIZ_STATUS_OFFSET);
-			if (val != 1) {
-				mdelay(100);
-				continue;
-			}
-		}
-		if (val != 1) {
+
+		err = clock_wiz_busy(clock, i, 100, 100);
```
Comment (larry9523) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c:860]: nit, will the log message looks like
"
clockwiz(%d) is (%u) busy after %d ms
clockwiz MMCM/PLL did not lock after %d ms, restoring the original configuration
"
Are the two "ms", %d and MMCM/PLL same thing? If so, can we skip something from the second log line?

Comment (xdavidz) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/clock.c:860]: yeah. fixed.

**Comment by xdavidz** [comment]
> Overall, this looks good.
> The tables, frequency_table[] and acap_frequency_table[], shared struct xclmgmt_ocl_clockwiz, but actually they store very different data (But column 1 is the same). This might make people confusing if he/she doesn't familiar with the code.
> There are separate functions to handle the difference, and XOCL_DSA_IS_VERSAL() is used again in the code.
> Should we create a separate subdev? Should we select the working table in probe?
> This is not a change request. It doesn't impact the functionality. The code is still clear and easy to understand.

It's a bit harder to split the clock subdev into two clock subdevs, literary they are sharing the clock subdev infra.

It would be nice to have a working table during probe. I would like to refactor the code later for both versal clock and untra scale clock.

The IS_VERSAL should be removed once we are fully data-driven. I think the zocl_ert design would be a good example to guide the refactoring work. It won't be a small code change though. I have created a sub-task for the clock refactoring work. http://jira.xilinx.com/browse/XRT-563

**Pull request merged**